### 🛒 Retail Demand Forecasting

#### Phase 1: Data Understanding

##### 1. Objective

The objective of this notebook is to understand the available datasets, identify their structure, analyze data quality, and understand how the datasets are related before performing exploratory data analysis and feature engineering.# New Section

In [2]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

pd.set_option("display.max_columns",None)
pd.set_option("display.max_rows",100)


#### Load Data

In [56]:
train = pd.read_csv("/content/drive/MyDrive/Retail-Demand-Forecasting/Data/Raw/train.csv")
test = pd.read_csv("/content/drive/MyDrive/Retail-Demand-Forecasting/Data/Raw/test.csv")
items = pd.read_csv("/content/drive/MyDrive/Retail-Demand-Forecasting/Data/Raw/items.csv")
stores = pd.read_csv("/content/drive/MyDrive/Retail-Demand-Forecasting/Data/Raw/stores.csv")
oil = pd.read_csv("/content/drive/MyDrive/Retail-Demand-Forecasting/Data/Raw/oil.csv")
transactions = pd.read_csv("/content/drive/MyDrive/Retail-Demand-Forecasting/Data/Raw/transactions.csv")
holidays = pd.read_csv("/content/drive/MyDrive/Retail-Demand-Forecasting/Data/Raw/holidays_events.csv")

#### Dataset summary

In [33]:
datasets = {"train":train,
            "test" : test,
            "items" :items,
            "stores" : stores,
            "oil" : oil,
            "transactions" : transactions,
            "holidays" :holidays,
            }

summary = pd.DataFrame({
    "Dataset": datasets.keys(),
    "Rows": [df.shape[0] for df in datasets.values()],
    "Columns": [df.shape[1] for df in datasets.values()],
    "Missing Values": [df.isnull().sum().sum() for df in datasets.values()],
    "Missing %": [round((df.isnull().sum().sum() / df.size) * 100, 2) for df in datasets.values()],
    "Memory Usage (MB)": [round(df.memory_usage(deep=True).sum() / (1024**2), 2) for df in datasets.values()]
})

summary


,Dataset,Rows,Columns,Missing Values,Missing %,Memory Usage (MB)
0,train,125497040,6,21657651,2.88,15117.16
1,test,3370464,5,0,0.00,270.00
2,items,4100,4,0,0.00,0.32
3,stores,54,5,0,0.00,0.01
4,oil,1218,2,43,1.77,0.08
5,transactions,83488,3,0,0.00,5.97
6,holidays,350,6,0,0.00,0.10


In [34]:
summary

,Dataset,Rows,Columns,Missing Values,Missing %,Memory Usage (MB)
0,train,125497040,6,21657651,2.88,15117.16
1,test,3370464,5,0,0.00,270.00
2,items,4100,4,0,0.00,0.32
3,stores,54,5,0,0.00,0.01
4,oil,1218,2,43,1.77,0.08
5,transactions,83488,3,0,0.00,5.97
6,holidays,350,6,0,0.00,0.10


### Key Observations
* The train dataset is the largest dataset, containing over 125 million records, indicating a large-scale retail demand forecasting problem.
Approximately 2.88% of the values are missing in the train dataset.
* Further investigation is required to identify the affected columns before deciding on an appropriate treatment strategy.
* The oil dataset contains 43 missing values (1.77%). Due to its relatively small size, dropping records may not be the best approach, and suitable imputation methods should be evaluated.
* All remaining datasets are complete and do not contain missing values.
* The train dataset consumes approximately 15 GB of memory, suggesting that memory-efficient processing techniques may be required during model development.

In [57]:
from IPython.display import display

def observations(df):
    print("=" * 80)
    print("First 5 Rows")
    display(df.head())

    print("=" * 80)
    print("Shape")
    display(df.shape)

    print("=" * 80)
    print("Data Information")
    df.info()

    print("=" * 80)
    print("Descriptive Statistics")
    display(df.describe(include="all"))

    print("=" * 80)
    print("Unique Values")
    display(df.nunique())

    print("=" * 80)
    print("Data Types")
    display(df.dtypes)

    print("=" * 80)
    print("Missing Values")
    display(df.isnull().sum())

    return df

### Train Dataset Analysis

In [50]:
observations(train)

First 5 Rows


,id,date,store_nbr,item_nbr,unit_sales,onpromotion
0,0,2013-01-01,25,103665,7.0,NaN
1,1,2013-01-01,25,105574,1.0,NaN
2,2,2013-01-01,25,105575,2.0,NaN
3,3,2013-01-01,25,108079,1.0,NaN
4,4,2013-01-01,25,108701,1.0,NaN


Shape


(125497040, 6)

Data Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125497040 entries, 0 to 125497039
Data columns (total 6 columns):
 #   Column       Dtype  
---  ------       -----  
 0   id           int64  
 1   date         object 
 2   store_nbr    int64  
 3   item_nbr     int64  
 4   unit_sales   float64
 5   onpromotion  object 
dtypes: float64(1), int64(3), object(2)
memory usage: 5.6+ GB
Descriptive Statistics


,id,date,store_nbr,item_nbr,unit_sales,onpromotion
count,1.254970e+08,125497040,1.254970e+08,1.254970e+08,1.254970e+08,103839389
unique,NaN,1684,NaN,NaN,NaN,2
top,NaN,2017-07-01,NaN,NaN,NaN,False
freq,NaN,118194,NaN,NaN,NaN,96028767
mean,6.274852e+07,NaN,2.746458e+01,9.727692e+05,8.554865e+00,NaN
std,3.622788e+07,NaN,1.633051e+01,5.205336e+05,2.360515e+01,NaN
min,0.000000e+00,NaN,1.000000e+00,9.699500e+04,-1.537200e+04,NaN
25%,3.137426e+07,NaN,1.200000e+01,5.223830e+05,2.000000e+00,NaN
50%,6.274852e+07,NaN,2.800000e+01,9.595000e+05,4.000000e+00,NaN
75%,9.412278e+07,NaN,4.300000e+01,1.354380e+06,9.000000e+00,NaN


Unique Values


,0
id,125497040
date,1684
store_nbr,54
item_nbr,4036
unit_sales,258474
onpromotion,2


Data Types


,0
id,int64
date,object
store_nbr,int64
item_nbr,int64
unit_sales,float64
onpromotion,object


Missing Values


,0
id,0
date,0
store_nbr,0
item_nbr,0
unit_sales,0
onpromotion,21657651


,id,date,store_nbr,item_nbr,unit_sales,onpromotion
0,0,2013-01-01,25,103665,7.0,NaN
1,1,2013-01-01,25,105574,1.0,NaN
2,2,2013-01-01,25,105575,2.0,NaN
3,3,2013-01-01,25,108079,1.0,NaN
4,4,2013-01-01,25,108701,1.0,NaN
...,...,...,...,...,...,...
125497035,125497035,2017-08-15,54,2089339,4.0,False
125497036,125497036,2017-08-15,54,2106464,1.0,True
125497037,125497037,2017-08-15,54,2110456,192.0,False
125497038,125497038,2017-08-15,54,2113914,198.0,True


#### Key Observations

- The train dataset contains **125.5 million records**, making it the largest dataset in the project.
- The dataset contains **6 columns**, including the target variable (`unit_sales`) and promotion information (`onpromotion`).
- The `date` column is currently stored as an **object** and should be converted to **datetime** for time-series analysis.
- There are **54 unique stores** and **4,036 unique products**, indicating forecasting at the Store–Item level.
- The `onpromotion` column contains approximately **17.26% missing values**, while all other columns are complete.
- The dataset occupies approximately **5.6 GB** of memory, which may require memory-efficient processing during feature engineering and model training.

#### Business Interpretation

- Historical sales data will be used to train forecasting models.
- Promotion is expected to be one of the most important drivers of product demand.
- Since only the promotion column contains missing values, data quality issues appear to be limited.
- The large dataset size suggests that scalable preprocessing and modeling techniques will be important.

### Test Dataset Analysis

In [58]:
observations(test)


First 5 Rows


,id,date,store_nbr,item_nbr,onpromotion
0,125497040,2017-08-16,1,96995,False
1,125497041,2017-08-16,1,99197,False
2,125497042,2017-08-16,1,103501,False
3,125497043,2017-08-16,1,103520,False
4,125497044,2017-08-16,1,103665,False


Shape


(3370464, 5)

Data Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3370464 entries, 0 to 3370463
Data columns (total 5 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   id           int64 
 1   date         object
 2   store_nbr    int64 
 3   item_nbr     int64 
 4   onpromotion  bool  
dtypes: bool(1), int64(3), object(1)
memory usage: 106.1+ MB
Descriptive Statistics


,id,date,store_nbr,item_nbr,onpromotion
count,3.370464e+06,3370464,3.370464e+06,3.370464e+06,3370464
unique,NaN,16,NaN,NaN,2
top,NaN,2017-08-16,NaN,NaN,False
freq,NaN,210654,NaN,NaN,3171867
mean,1.271823e+08,NaN,2.750000e+01,1.244798e+06,NaN
std,9.729693e+05,NaN,1.558579e+01,5.898362e+05,NaN
min,1.254970e+08,NaN,1.000000e+00,9.699500e+04,NaN
25%,1.263397e+08,NaN,1.400000e+01,8.053210e+05,NaN
50%,1.271823e+08,NaN,2.750000e+01,1.294665e+06,NaN
75%,1.280249e+08,NaN,4.100000e+01,1.730015e+06,NaN


Unique Values


,0
id,3370464
date,16
store_nbr,54
item_nbr,3901
onpromotion,2


Data Types


,0
id,int64
date,object
store_nbr,int64
item_nbr,int64
onpromotion,bool


Missing Values


,0
id,0
date,0
store_nbr,0
item_nbr,0
onpromotion,0


,id,date,store_nbr,item_nbr,onpromotion
0,125497040,2017-08-16,1,96995,False
1,125497041,2017-08-16,1,99197,False
2,125497042,2017-08-16,1,103501,False
3,125497043,2017-08-16,1,103520,False
4,125497044,2017-08-16,1,103665,False
...,...,...,...,...,...
3370459,128867499,2017-08-31,54,2132163,False
3370460,128867500,2017-08-31,54,2132318,False
3370461,128867501,2017-08-31,54,2132945,False
3370462,128867502,2017-08-31,54,2132957,False


### key Observations

* The test dataset contains 3,901 products compared to 4,036 products in the training dataset.

* This suggests that some products are absent from the forecasting period. Further investigation is required to determine whether these products were discontinued, seasonal, or intentionally excluded from the prediction dataset.

* The onpromotion column is stored as a boolean variable in the test dataset, whereas it is an object in the training dataset due to the presence of missing values in the training data.

In [59]:
observations(oil)

First 5 Rows


,date,dcoilwtico
0,2013-01-01,NaN
1,2013-01-02,93.14
2,2013-01-03,92.97
3,2013-01-04,93.12
4,2013-01-07,93.20


Shape


(1218, 2)

Data Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1218 entries, 0 to 1217
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        1218 non-null   object 
 1   dcoilwtico  1175 non-null   float64
dtypes: float64(1), object(1)
memory usage: 19.2+ KB
Descriptive Statistics


,date,dcoilwtico
count,1218,1175.000000
unique,1218,NaN
top,2017-08-31,NaN
freq,1,NaN
mean,NaN,67.714366
std,NaN,25.630476
min,NaN,26.190000
25%,NaN,46.405000
50%,NaN,53.190000
75%,NaN,95.660000


Unique Values


,0
date,1218
dcoilwtico,998


Data Types


,0
date,object
dcoilwtico,float64


Missing Values


,0
date,0
dcoilwtico,43


,date,dcoilwtico
0,2013-01-01,NaN
1,2013-01-02,93.14
2,2013-01-03,92.97
3,2013-01-04,93.12
4,2013-01-07,93.20
...,...,...
1213,2017-08-25,47.65
1214,2017-08-28,46.40
1215,2017-08-29,46.46
1216,2017-08-30,45.96


In [68]:
store = observations(stores)


First 5 Rows


,store_nbr,city,state,type,cluster
0,1,Quito,Pichincha,D,13
1,2,Quito,Pichincha,D,13
2,3,Quito,Pichincha,D,8
3,4,Quito,Pichincha,D,9
4,5,Santo Domingo,Santo Domingo de los Tsachilas,D,4


Shape


(54, 5)

Data Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   store_nbr  54 non-null     int64 
 1   city       54 non-null     object
 2   state      54 non-null     object
 3   type       54 non-null     object
 4   cluster    54 non-null     int64 
dtypes: int64(2), object(3)
memory usage: 2.2+ KB
Descriptive Statistics


,store_nbr,city,state,type,cluster
count,54.000000,54,54,54,54.000000
unique,NaN,22,16,5,NaN
top,NaN,Quito,Pichincha,D,NaN
freq,NaN,18,19,18,NaN
mean,27.500000,NaN,NaN,NaN,8.481481
std,15.732133,NaN,NaN,NaN,4.693395
min,1.000000,NaN,NaN,NaN,1.000000
25%,14.250000,NaN,NaN,NaN,4.000000
50%,27.500000,NaN,NaN,NaN,8.500000
75%,40.750000,NaN,NaN,NaN,13.000000


Unique Values


,0
store_nbr,54
city,22
state,16
type,5
cluster,17


Data Types


,0
store_nbr,int64
city,object
state,object
type,object
cluster,int64


Missing Values


,0
store_nbr,0
city,0
state,0
type,0
cluster,0


### Key Observations (stores)

- The dataset contains metadata for all 54 stores used in the project.
- Stores are distributed across 22 cities and 16 states, providing geographical information.
- Stores are categorized into 5 different types and 17 predefined clusters.
- No missing values are present in the dataset.
- The dataset can be merged with the sales data using `store_nbr` to enrich the forecasting model with store-level features.

### Business Interpretation

- Geographic attributes such as city and state can capture regional demand variations.
- Store type and cluster may explain differences in customer behavior and store characteristics.
- These attributes are expected to improve forecasting by incorporating store-specific information beyond historical sales.

In [70]:
observations(transactions)


First 5 Rows


,date,store_nbr,transactions
0,2013-01-01,25,770
1,2013-01-02,1,2111
2,2013-01-02,2,2358
3,2013-01-02,3,3487
4,2013-01-02,4,1922


Shape


(83488, 3)

Data Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83488 entries, 0 to 83487
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          83488 non-null  object
 1   store_nbr     83488 non-null  int64 
 2   transactions  83488 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 1.9+ MB
Descriptive Statistics


,date,store_nbr,transactions
count,83488,83488.000000,83488.000000
unique,1682,NaN,NaN
top,2017-07-03,NaN,NaN
freq,54,NaN,NaN
mean,NaN,26.939237,1694.602158
std,NaN,15.608204,963.286644
min,NaN,1.000000,5.000000
25%,NaN,13.000000,1046.000000
50%,NaN,27.000000,1393.000000
75%,NaN,40.000000,2079.000000


Unique Values


,0
date,1682
store_nbr,54
transactions,4993


Data Types


,0
date,object
store_nbr,int64
transactions,int64


Missing Values


,0
date,0
store_nbr,0
transactions,0


,date,store_nbr,transactions
0,2013-01-01,25,770
1,2013-01-02,1,2111
2,2013-01-02,2,2358
3,2013-01-02,3,3487
4,2013-01-02,4,1922
...,...,...,...
83483,2017-08-15,50,2804
83484,2017-08-15,51,1573
83485,2017-08-15,52,2255
83486,2017-08-15,53,932


### Business Interpretation (Transactions)

- The transactions dataset captures daily customer activity at each store.
- Customer traffic is expected to have a strong relationship with sales volume.
- Transaction counts can be used as an external feature to improve forecasting performance.
- The dataset can be merged with the sales data using both `date` and `store_nbr`.

In [65]:
observations(items)


First 5 Rows


,item_nbr,family,class,perishable
0,96995,GROCERY I,1093,0
1,99197,GROCERY I,1067,0
2,103501,CLEANING,3008,0
3,103520,GROCERY I,1028,0
4,103665,BREAD/BAKERY,2712,1


Shape


(4100, 4)

Data Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4100 entries, 0 to 4099
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   item_nbr    4100 non-null   int64 
 1   family      4100 non-null   object
 2   class       4100 non-null   int64 
 3   perishable  4100 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 128.3+ KB
Descriptive Statistics


,item_nbr,family,class,perishable
count,4.100000e+03,4100,4100.0000,4100.000000
unique,NaN,33,NaN,NaN
top,NaN,GROCERY I,NaN,NaN
freq,NaN,1334,NaN,NaN
mean,1.251436e+06,NaN,2169.6500,0.240488
std,5.876872e+05,NaN,1484.9109,0.427432
min,9.699500e+04,NaN,1002.0000,0.000000
25%,8.181108e+05,NaN,1068.0000,0.000000
50%,1.306198e+06,NaN,2004.0000,0.000000
75%,1.904918e+06,NaN,2990.5000,0.000000


Unique Values


,0
item_nbr,4100
family,33
class,337
perishable,2


Data Types


,0
item_nbr,int64
family,object
class,int64
perishable,int64


Missing Values


,0
item_nbr,0
family,0
class,0
perishable,0


,item_nbr,family,class,perishable
0,96995,GROCERY I,1093,0
1,99197,GROCERY I,1067,0
2,103501,CLEANING,3008,0
3,103520,GROCERY I,1028,0
4,103665,BREAD/BAKERY,2712,1
...,...,...,...,...
4095,2132318,GROCERY I,1002,0
4096,2132945,GROCERY I,1026,0
4097,2132957,GROCERY I,1068,0
4098,2134058,BEVERAGES,1124,0


### Business Interpretation (Items)

- Product family captures broad product categories and can help identify category-specific demand patterns.
- Product class provides a more granular categorization within each family.
- The perishable flag is expected to be an important predictor, as demand patterns for perishable products differ significantly from non-perishable products.
- Product attributes can be merged with the sales dataset using `item_nbr` to enrich the forecasting model.

In [67]:
holiday = observations(holidays)
holiday

First 5 Rows


,date,type,locale,locale_name,description,transferred
0,2012-03-02,Holiday,Local,Manta,Fundacion de Manta,False
1,2012-04-01,Holiday,Regional,Cotopaxi,Provincializacion de Cotopaxi,False
2,2012-04-12,Holiday,Local,Cuenca,Fundacion de Cuenca,False
3,2012-04-14,Holiday,Local,Libertad,Cantonizacion de Libertad,False
4,2012-04-21,Holiday,Local,Riobamba,Cantonizacion de Riobamba,False


Shape


(350, 6)

Data Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   date         350 non-null    object
 1   type         350 non-null    object
 2   locale       350 non-null    object
 3   locale_name  350 non-null    object
 4   description  350 non-null    object
 5   transferred  350 non-null    bool  
dtypes: bool(1), object(5)
memory usage: 14.1+ KB
Descriptive Statistics


,date,type,locale,locale_name,description,transferred
count,350,350,350,350,350,350
unique,312,6,3,24,103,2
top,2014-06-25,Holiday,National,Ecuador,Carnaval,False
freq,4,221,174,174,10,338


Unique Values


,0
date,312
type,6
locale,3
locale_name,24
description,103
transferred,2


Data Types


,0
date,object
type,object
locale,object
locale_name,object
description,object
transferred,bool


Missing Values


,0
date,0
type,0
locale,0
locale_name,0
description,0
transferred,0


,date,type,locale,locale_name,description,transferred
0,2012-03-02,Holiday,Local,Manta,Fundacion de Manta,False
1,2012-04-01,Holiday,Regional,Cotopaxi,Provincializacion de Cotopaxi,False
2,2012-04-12,Holiday,Local,Cuenca,Fundacion de Cuenca,False
3,2012-04-14,Holiday,Local,Libertad,Cantonizacion de Libertad,False
4,2012-04-21,Holiday,Local,Riobamba,Cantonizacion de Riobamba,False
...,...,...,...,...,...,...
345,2017-12-22,Additional,National,Ecuador,Navidad-3,False
346,2017-12-23,Additional,National,Ecuador,Navidad-2,False
347,2017-12-24,Additional,National,Ecuador,Navidad-1,False
348,2017-12-25,Holiday,National,Ecuador,Navidad,False


### Business Interpretation (Holiday)

- Holidays are expected to have a significant impact on retail demand.
- National holidays affect all stores, while regional and local holidays influence only specific geographical locations.
- Different event types (Holiday, Bridge, Additional, etc.) may have varying effects on customer purchasing behavior.
- The holiday dataset can be merged with the sales data using the date field and further filtered using store location information.

# Overall Dataset Understanding

The Corporación Favorita dataset consists of one primary sales dataset and multiple supporting datasets that provide additional business context for demand forecasting.

The **train** dataset contains historical sales recorded at the **Store × Item × Date** level and serves as the primary dataset for model training. The **test** dataset contains the future dates for which sales predictions need to be generated.

The remaining datasets provide valuable external information that can improve forecasting performance:

- **Items** dataset provides product attributes such as family, class, and perishability.
- **Stores** dataset provides geographical and operational information including city, state, store type, and cluster.
- **Transactions** dataset captures daily customer traffic at the store level.
- **Oil** dataset provides daily crude oil prices, which may reflect macroeconomic conditions in Ecuador.
- **Holidays** dataset contains national, regional, and local holidays that may influence customer purchasing behavior.
- **Sample Submission** dataset defines the expected output format for Kaggle submissions.

## Key Learnings

From the dataset understanding phase, the following hypotheses have been identified:

- Promotional activities are likely to influence product demand.
- Holidays may create temporary spikes or dips in sales.
- Customer transactions may serve as a strong indicator of sales volume.
- Product characteristics such as family and perishability may affect demand patterns.
- Store location, type, and cluster may explain regional variations in sales.
- Oil prices may indirectly impact retail demand by reflecting broader economic conditions in Ecuador.

## Data Quality Summary

- The train dataset contains missing values only in the `onpromotion` column.
- The oil dataset contains a small number of missing oil prices.
- All other datasets are complete and contain no missing values.
- Date columns will be converted to datetime format before analysis.

## Next Steps

The next phase of the project will focus on Exploratory Data Analysis (EDA), where these hypotheses will be validated by analyzing:

- Sales trends over time
- Seasonality and cyclic patterns
- Promotion impact on sales
- Holiday impact on demand
- Product family performance
- Store-wise and regional sales patterns
- Relationship between oil prices and sales
- Relationship between customer transactions and sales

The insights obtained from EDA will guide feature engineering and model selection for the forecasting pipeline.